### PR TITLE
Prevent IO exceptions

### LINKED ISSUE #1780: Language setting not persistent
Hello,

I'm french but I play ED in english. I use VoiceAttack with EDDI as plugin. I can set the language to english and it works for a while. But that setting randomly falls back to french. 
It does not happen everytime.

When it happens, I need to open the EDDI standalone and change the language back to english. The combo is blank at that time (cf screenshot). Like it's unable to load the config file and falls back to default.
language settings blank

I used VA/EDDI a year ago with Windows 7 and all was fine. I then stopped playing for a few month.
I now use Windows 10.

## EDDI version in which issue is found
3.5.2

## VoiceAttack version in which issue is found (as applicable)
1.8.5

## Steps to reproduce
Set language in EDDI standalone. Use in voice attack.
On a subsequent session, the language setting in EDDI is blank.
It's random.
I could reproduce today stopping/starting voice attack several times. After the 3rd or 4th, it happened.

## Expected
Language setting kept

## Observed
Language setting randomly blank in EDDI

## Investigation
Tried running both EDDI and VA as admin. Nothing changed.

### PR DESCRIPTION
**Pull request by Tkael**
This PR is designed to prevent issues like:
- https://rollbar.com/EDCD/EDDI/items/17558/
- https://rollbar.com/EDCD/EDDI/items/16979/
- https://rollbar.com/EDCD/EDDI/items/17609/
- https://rollbar.com/EDCD/EDDI/items/17607/
- https://rollbar.com/EDCD/EDDI/items/17583/

It moves file locking from each individual config file into the `Files` class and if an IO exception occurs it waits and retries several times before completely failing.

May resolve #1780.

### CODE REVIEW

**Comment by richardbuckle** [comment]
This one is pervasive. I need to consider it in depth.

**Comment by richardbuckle** [comment]
Anything involving the file system, locking and IO exceptions is for me a trigger for a deeper code review than usual. #NotMyFirstRodeo :)

File: Utilities/Files.cs
Comment (richardbuckle) [inline Utilities/Files.cs]: Wont these loop infinitely, since attempts is not decremented or zeroed?

File: Utilities/Files.cs
Comment (richardbuckle) [inline Utilities/Files.cs]: Likewise I think these will loop infinitely, as attempts is not decremented or zeroed.

**Review by richardbuckle — CHANGES_REQUESTED** [review]
Unless I am misunderstanding something, I think I have found a flaw. @GioVAX could you have a look please? A fresh pair of eyes might be good here.

File: Utilities/Files.cs
Comment (Tkael) [reply Utilities/Files.cs]: We break out of the loop, regardless of whether `attempts` decrements to zero, using `break;` at line 83.

File: Utilities/Files.cs
Comment (Tkael) [reply Utilities/Files.cs]: We break out of the loop, regardless of whether `attempts` decrements to zero, using `break;` at line 157.

File: Utilities/Files.cs
Comment (GioVAX) [reply Utilities/Files.cs]: Maybe for clarity, the `break` could be moved to line 37?
So it gets executed only if the read statement doesn't throw.

File: Utilities/Files.cs
Comment (GioVAX) [inline Utilities/Files.cs]: Just curious: are you sure that `Unlock(filename)` will get executed if action throws?

File: Utilities/Files.cs
Comment (GioVAX) [inline Utilities/Files.cs]: NIT - this class is generic. I'd use `lockname` instead of `filename`

File: Utilities/Files.cs
Comment (GioVAX) [inline Utilities/Files.cs]: I am surely missing something here: I assumed that we use locks to prevent concurrent write access to files, but if we increment the refcount and return the lock, we are not blocking the second access.

File: Utilities/Files.cs
Comment (richardbuckle) [reply Utilities/Files.cs]: No, that would change the logic.

I think the structure is very confusing and fragile. 
1. Only `catch (IOException ex) when ((ex.HResult & 0x0000FFFF) == 32) // Sharing violation` with `ignoreMissing == false` has a path that decrements `attempts`. 
2. All other catch statements and the happy path lead to a `break`.

I think the control logic needs a bit of a re-think here.

File: Utilities/Files.cs
Comment (Tkael) [reply Utilities/Files.cs]: Why would incrementing the ref count unblock anything? The same lock object is being used.

File: Utilities/Files.cs
Comment (Tkael) [reply Utilities/Files.cs]: It's not a public class. It's internal to Files.cs. While it could be extracted as a generic class, that is not the way we are currently using it.

File: Utilities/Files.cs
Comment (Tkael) [reply Utilities/Files.cs]: Good point. I'll wrap it with try-finally to ensure that `Unlock` will execute regardless of any throw.

File: Utilities/Files.cs
Comment (Tkael) [reply Utilities/Files.cs]: I'll go ahead and move it to a public class so that it can be used a general utility.

File: Utilities/Files.cs
Comment (Tkael) [reply Utilities/Files.cs]: Moved to LockManager.cs.

Comment (GioVAX) [reply Utilities/Files.cs]: You are right, I missed the `lock(` in front of the call to `GetLock`.

Do we then need refcount?
Every lock should be used only by one thread at the time, right?

Comment (GioVAX) [reply Utilities/Files.cs]: Upon further thinking about this, I see why we need ref counting.
Please, forget my previous comment :)

**Comment by Tkael** [comment]
Closing to rebase and resolve a minor merge conflict.

**Pull request closed without merging**
